add comma operator to the deprecated list

wilzbach opened the pull request:
As we are finally getting rid of the comma operator, how about adding it to the "Future" list. Once the [PR](https://github.com/dlang/dmd/pull/5737) is merged, more infos can be added?
Ping @Geod24 @yebblies

Hackerpilot commented on deprecate.dd:
```diff
@@ -55,6 +56,32 @@ $(SPEC_S Deprecated Features,
+    $(P Usage of the comma operator outside of for-loops should be forbidden.
+    )
+$(H4 Rationale)
+    $(P The comma operator leads to a lot of false friends,
```
Maybe rephrase as "unintended behavior"

Hackerpilot commented on deprecate.dd:
```diff
@@ -55,6 +56,32 @@ $(SPEC_S Deprecated Features,
 
 
+$(H3 $(DEPNAME Comma operator))
+    $(P The comma operator (`,`) allows to execute multiple expressions and
```
"allows executing multiple"

Hackerpilot commented on deprecate.dd:
```diff
@@ -55,6 +56,32 @@ $(SPEC_S Deprecated Features,
+    $(P The comma operator leads to a lot of false friends,
+        (e.g. `synchronized (lockA, lockB)` isn't currently implemented, but still
+        compiles "thanks" to the comma operator).
+        Moreover it has a tiny (mostly accidental) usage
```
Maybe "Moreover it is not commonly used and it blocks the ability to implement tuples"

wilzbach commented:
Thanks @Hackerpilot - addressed and rebased.

yebblies commented on deprecate.dd:
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
 
+$(H3 $(DEPNAME Comma operator))
+    $(P The comma operator (`,`) allows executing multiple expressions and
+        discards them except for the last which is returned.
```
The expressions aren't discarded, only their results.

yebblies commented on deprecate.dd:
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
+        ---
+
+        It's also common to use the comma operator in for-loop increment
+        statements to allow multiple statements.
```
multiple expressions, not statements.

yebblies commented on deprecate.dd:
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Usage of the comma operator outside of for-loops should be forbidden.
```
This is not the path we're going with.  Please take a look at the current PRs.

yebblies commented on deprecate.dd:
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
+$(H4 Rationale)
+    $(P The comma operator leads to unintended behavior,
+        (e.g. `synchronized (lockA, lockB)` isn't currently implemented, but still
+        compiles "thanks" to the comma operator).
```
This is not the best example, since it's actually a parser bug.  There are plenty of more common examples.

Hackerpilot replied:
No it's not. The grammar says:
```
SynchronizedStatement:
    synchronized ( Expression ) ScopeStatement
Expression:
    CommaExpression
```
It's perfectly valid for a D parser to accept this code.

yebblies replied:
I feel like we've had this discussion before and you probably know what I mean.  Spec bug, parser bug, whatever.  The point is that even with comma expressions still allowed in the language we could avoid accepting this bug by changing the parser.

wilzbach replied:
> There are plenty of more common examples.

I went through the github issue and found those:

```
writeln( 6, mixin("7,8"), 9 ); // 6, 8, 9
vec v = (0, 0, 3); // 3

int b = 2;
if (a == 1, b == 2) {
  // will always be reached
}
```

do you have better ones in mind? Just pick an example ;-)

wilzbach commented on deprecate.dd:
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Using the result of the comma operator should be forbidden.
```
@yebblies did I formulate the action correctly?
Isn't it still planned to deprecate comma operator on the long run and replace it with tuples?

mathias-lang-sociomantic commented:
Regarding the 'needs approval', I believe this change already have been approved: https://forum.dlang.org/post/nh1luh$20ds$1@digitalmars.com

mathias-lang-sociomantic commented on deprecate.dd:
```diff
@@ -10,6 +10,7 @@ $(SPEC_S Deprecated Features,
 
     $(TABLE2 Deprecated Features,
         $(THEAD Feature,                                                          Spec,  Dep,    Error,  Gone)
+        $(TROW $(DEPLINK Comma operator),                                         future, &nbsp;, &nbsp;, &nbsp;)
```
"Using the result of a comma expression"
We can extend it later, but so far that's what we are discussing.

wilzbach commented:
@mathias-lang-sociomantic I updated the title and added a list of unintended behaviors with the comma operator - I hope everyone can find his favorite bug in there ;-)

wilzbach commented on deprecate.dd:
```diff
@@ -55,6 +56,48 @@ $(SPEC_S Deprecated Features,
+    $(P The comma operator leads to unintended behavior (see below for a selection)
+        Moreover it is not commonly used and it blocks the ability to implement tuples.
+
+        A selection of problems through the accidental use of the comma operator:
```
is this line redundant?

wilzbach commented:
So if this is already approved, we can move forward?
Any other comments?

yebblies commented on deprecate.dd:
```diff
@@ -55,6 +56,48 @@ $(SPEC_S Deprecated Features,
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Using the result of the comma operator should be forbidden.
```
This should be a description of how to fix code - ie replace with a lambda or statement.  Check the other items on this page for other examples of what the Corrective Action section is for.

yebblies commented:
Apart from that looks ok.

wilzbach commented:
> Apart from that looks ok.

rebased.

wilzbach commented:
@yebblies @mathias-lang-sociomantic more nitpicks or are we ready to go?

yebblies commented:
Yup LGTM.  I don't have merge rights here or I'd merge it.
